modules/ghlint-rules/src/main/kotlin/net/twisterrob/ghlint/rules/NeverExpressionRule.kt

```diff
@@ -0,0 +1,72 @@
+package net.twisterrob.ghlint.rules
+
+import net.twisterrob.ghlint.model.WorkflowStep
+import net.twisterrob.ghlint.rule.Example
+import net.twisterrob.ghlint.rule.Issue
+import net.twisterrob.ghlint.rule.Reporting
+import net.twisterrob.ghlint.rule.report
+import net.twisterrob.ghlint.rule.visitor.VisitorRule
+import net.twisterrob.ghlint.rule.visitor.WorkflowVisitor
+
+public class NeverExpressionRule : VisitorRule, WorkflowVisitor {
+
+    override val issues : List<Issue> = listOf(NeverExpression)
+
+    private companion object {
+        const val uses = "actions/checkout@\${{ github.sha }}"
+
+        val NeverExpression = Issue(
+            id = "NeverExpression",
+            title = "Expressions should not be used in uses field.",
+            description = """
+            > GitHub Action Expressions can be used to programmatically set environment variables in workflow files.
+            > -- [About expressions](https://docs.github.com/en/actions/learn-github-actions/expressions)
+            
+            However, they cannot be used within the uses field. GitHub will report an error - "The workflow is not valid: A template expression is not allowed in this context"
+        """.trimIndent(),
+            compliant = listOf(
+                Example(
+                    explanation = "GitHub Expression not used within the uses field.",
+                    path = "ci-build.yml",
+                    content = """
+						on: push
+						jobs:
+						  example:
+						    runs-on: ubuntu-latest
+						    steps:
+						      - uses: actions/checkout@v4"
+					""".trimIndent(),
+                ),
+            ),
+            nonCompliant = listOf(
+                Example(
+                    explanation = "GitHub Expression used within the uses field.",
+                    path = "ci-build.yml",
+                    content = """
+						on: push
+						jobs:
+						  example:
+						    runs-on: ubuntu-latest
+						    steps:
+						      - uses: $uses
+					""".trimIndent(),
+                ),
+            )
+        )
+    }
+
+    override fun visitWorkflowUsesStep(reporting: Reporting, step: WorkflowStep.Uses) {
+        super.visitWorkflowUsesStep(reporting, step)
+
+        if (step.uses.uses.containsGitHubExpression()) {
+            reporting.report(NeverExpression, step) {
+                "$it uses a GitHub expression in the uses field."
```

Comment:
any chance for a different wording without the double uses? If any other option you can come up with feels awkward, then we can always fall back to:

```suggestion
                "${it} uses a GitHub expression in the `uses` field."
```

Reply:
Will have a think - btw, My IDEA was stating those brackets were redundant with a warning

Reply:
Yeah, I know, that's the default setting, because it's not _needed_, but it's similar to how we do `"${x}"` in shell everywhere, because if you forget, you might have a problem, better be safe than saving 2 characters sometimes.

Ideally this disables it: https://github.com/TWiStErRob/net.twisterrob.ghlint/pull/268/files#diff-bbf0d02ebcf6764cae92fa4045a2d6f15f8a479cd1aa2ab4d1fd435a3aa74bbdR9

and restart would pick it up.